[DM-4354] Document the feature enabling LWM2M to send device data over the C8y MQTT Service (#2812)
SoftwareAG/c8y-docs: content/protocol-integration/lwm2m-bundle/mqtt-service-integration.md

---
weight: 100
title: Send Device Data to Cumulocity’s MQTT Service
layout: redirect
---

Alternatively, LWM2M Service can send the device data to {{< product-c8y-iot >}}'s MQTT Service. 

> Once configured, you can easily access and process the data using your own microservice, giving you the flexibility 

Reviewer: Sentence is too long. Please break this up into two sentences.

Author: done

to handle the data in a way that suits your specific needs.

{{< c8y-admon-info >}}
If you are not familiar with MQTT, we recommend reviewing one of the many online resources available. For more 

> details, you can visit the [MQTT website](https://mqtt.org/mqtt-specification/).

Reviewer: I would rather point the customer here to our MQTT service documentation, not MQTT in general.

Author: done

To retrieve device data from the MQTT Service, 
you can create a custom microservice that suits your specific requirements. 
Once the microservice is developed, simply share it with your Operations Team, who will deploy it to the Platform for your tenant.

The microservice should be configured to subscribe to the `lwm2m/data` topic, 
where the {{< product-c8y-iot >}} sends the device data. After deployment, 
the microservice will authenticate using the tenant credentials associated with the deployment. 
For the microservice to have sufficient rights to connect and consume MQTT Service data, ensure that the following roles are assigned:

- `ROLE_INVENTORY_READ`
- `ROLE_NOTIFICATION_2_ADMIN`
- `ROLE_MQTT_SERVICE_ADMIN`

Please note that the microservice will only have access to data from the specified tenant and its subtenants. 
If you need to access data from a different tenant, a separate microservice must be deployed for that tenant.

If you need help developing your own microservice, 
you can refer to this [sample project](https://github.com/Cumulocity-IoT/lwm2m-devicedata-listener) for detailed guidance.

### MQTT Data Payload

> The data sent via MQTT uses the SenML-JSON format, 

Reviewer: Actually, that is not quite accurate: We send SenML-JSON that is encapsulated into a Cumulocity-specific wrapper JSON format.

Author: done

following the [RFC8428 specification](https://datatracker.ietf.org/doc/html/rfc8428). 
Keep in mind that LWM2M’s Time resource data is stored as a timestamp in seconds. 
Additionally, the Base Time may include fractional data and is represented in scientific notation.

Here’s an example of what the MQTT payload looks like when a device reports composite data for Object ID 
6 (Location) and Object ID 3303 (Temperature sensor):

```json
{
  "timestampDataReceived": "2025-01-15T18:03:52.881594523+01:00",
  "sourceOfData": "READ_RESPONSE",
  "deviceId": "65107",
  "tenantId": "management",
  "lwm2mEndpoint": "imei:test:device",
  "registrationId": "HKYbqsjcP7",

>   "senMLMessage": [

Reviewer: Only this part is SenMl. The other data is not.

---
weight: 100
title: Send Device Data to Cumulocity’s MQTT Service
layout: redirect
---

> This guide helps you configure the {{< product-c8y-iot >}} to send device data to Cumulocity's MQTT Service. 

Reviewer 2: ```suggestion
Alternatively, LWM2M Service can send the device data to {{< product-c8y-iot >}}'s MQTT Service. 
```

Author: Instead of LWM2M Service, can we use {{< product-c8y-iot >}} as it's common for most of the documentation when we want to mention the lwm2m agent?

Reviewer 2: {{< product-c8y-iot >}} placeholder is replaces with "Cumulocity" in our user doc.

Once configured, you can easily access and process the data using your own microservice, giving you the flexibility 
to handle the data in a way that suits your specific needs.

> {{< c8y-admon-important >}}
> To ensure a smooth setup, please make sure that both Pulsar and the MQTT Service are properly configured and 
> running before starting the {{< product-c8y-iot >}} configuration. Missing these prerequisites may result in excessive log 
> entries indicating that the {{< product-c8y-iot >}} cannot establish a connection to the MQTT Service.
> {{< /c8y-admon-important >}}

Reviewer 2: ```suggestion

```
Could you remove this? Setup should already be ready (done by OPS) before customer is using it.

### Why MQTT Service?

The MQTT Service, developed by Cumulocity, offers several important benefits:

- **Multi-tenancy support**: A single endpoint serves multiple tenants, simplifying management.
- **Identity management**: Services connected to the MQTT Service use Cumulocity tenant platform credentials.
- **Efficient communication**: Low traffic overhead, ensuring efficient data transfer.
- **WebSocket and TLS support**: Secure and flexible communication options.
- **Scalability**: The service scales horizontally to meet your growing needs.
- **Custom payload formats**: The service uses SenML-JSON, a widely used format for device measurement data.

Starting from release 2025, the {{< product-c8y-iot >}} can send device data to Cumulocity’s MQTT microservice, where the 
data is queued. This enables you to process and read the data via your custom service, with the added benefit 
of being able to leverage Cumulocity's integrated data analysis tools, should you choose to do so.

### Retrieving Device Data from the MQTT Service Using a Custom Microservice

To retrieve device data from the MQTT Service, 
you can create a custom microservice that suits your specific requirements. 

> Once the microservice is developed, simply share it with your Operations Team, who will deploy it to the Platform for your tenant.

Reviewer 2: Customer uploads from Administration application > Microservice UI.
Please refer to that doc with a link as well.

The microservice should be configured to subscribe to the `lwm2m/data` topic, 
where the {{< product-c8y-iot >}} sends the device data. After deployment, 
the microservice will authenticate using the tenant credentials associated with the deployment. 

> For the microservice to function correctly, ensure that the following roles are assigned:

Reviewer 2: ```suggestion
For the microservice to have sufficient rights to connect and consume MQTT Service data, ensure that the following roles are assigned:
```

---
weight: 100

> title: Send Device Data to Cumulocity’s MQTT Service

Reviewer 3: ```suggestion
title: Send device data to Cumulocity’s MQTT Service
```

layout: redirect
---

> Alternatively, LWM2M Service can send the device data to {{< product-c8y-iot >}}'s MQTT Service.

Reviewer 3: ```suggestion
Alternatively, the LWM2M service can send the device data to {{< product-c8y-iot >}}'s MQTT Service.
```

Once configured, you can access and process the data using your own microservice. 
This gives you the flexibility to handle the data in a way that suits your specific needs.

{{< c8y-admon-info >}}
If you're new to MQTT Service,

> we recommend reviewing the [MQTT Service documentation](/device-integration/mqtt/) for a better understanding.

Reviewer 3: ```suggestion
we recommend you to review the [MQTT Service documentation](/device-integration/mqtt/) for a better understanding.
```

{{< /c8y-admon-info >}}

> ### Why Choose MQTT Service?

Reviewer 3: ```suggestion
### Why choose MQTT Service?
```

Cumulocity's MQTT Service offers several key advantages that can help streamline your IoT operations:

- **Multi-tenancy support**: Manage multiple tenants with a single endpoint, simplifying operations.

> - **Identity management**: Easily manage service connections using Cumulocity platform credentials.

Reviewer 3: ```suggestion
- **Identity management**: Easily manage service connections using {{< product-c8y-iot >}} platform credentials.
```

- **Efficient communication**: Benefit from low traffic overhead for quick and efficient data transfer.
- **Secure communication**: Enjoy the flexibility of WebSocket and TLS support for secure communication.
- **Scalability**: Scale your operations seamlessly as your needs grow.
- **Custom payload formats**: Data is transmitted in JSON format, wrapping SenML-JSON for device-related information.

> Starting from release 2025, the LWM2M Service can send device data to Cumulocity's MQTT microservice, 

Reviewer 3: ```suggestion
Starting from release 2025, the LWM2M Service can send device data to {{< product-c8y-iot >}}'s MQTT microservice, 
```

where it gets queued. This allows you to process and access the data via your custom service 

> and take advantage of Cumulocity’s integrated data analysis tools, if needed.

Reviewer 3: ```suggestion
and take advantage of {{< product-c8y-iot >}}’s integrated data analysis tools, if needed.
```

> ### How to Retrieve Device Data from MQTT Service Using Your Custom Microservice

Reviewer 3: ```suggestion
### How to retrieve device data from MQTT Service using your custom microservice
```

You can retrieve device data from the MQTT Service 
by creating your own custom microservice tailored to your requirements. 
To learn how to develop a {{< product-c8y-iot >}} microservice, 

> refer to the [Microservice SDK guide](/microservice-sdk/).

Reviewer 3: ```suggestion
refer to [Microservice SDK](/microservice-sdk/).
```

Once your microservice is ready, deploy it to {{< product-c8y-iot >}} 
by following the steps outlined in [Managing microservices](/standard-tenant/ecosystem/#managing-microservices).

> Your microservice should subscribe to the `lwm2m/data` topic, where the LWM2M Service sends device data. 

Reviewer 3: ```suggestion
Your microservice should subscribe to the `lwm2m/data` topic, where the LWM2M service sends device data. 
```

After deployment, the microservice will authenticate using the tenant credentials associated with your deployment. 
To ensure the microservice has sufficient permissions to access the MQTT Service data, assign the following roles:

- `ROLE_INVENTORY_READ`
- `ROLE_NOTIFICATION_2_ADMIN`
- `ROLE_MQTT_SERVICE_ADMIN`

> Please note that the microservice will only have access to data from the specific tenant and its subtenants. 

Reviewer 3: ```suggestion
Note that the microservice will only have access to data from the specific tenant and its subtenants. 
```

If you need to access data from a different tenant, a separate microservice must be deployed for that tenant.

If you need help getting started, 
check out this [sample project](https://github.com/Cumulocity-IoT/lwm2m-devicedata-listener) for a detailed walkthrough.

> ### Understanding MQTT Data Payload

Reviewer 3: ```suggestion
### Understanding MQTT data payload
```

> The data transmitted via MQTT follows a custom Cumulocity JSON format, which includes:

Reviewer 3: ```suggestion
The data transmitted via MQTT follows a custom {{< product-c8y-iot >}} JSON format, which includes:
```

| Field                 | Description                                                                 |
|-----------------------|-----------------------------------------------------------------------------|

> | `timestampDataReceived` | Timestamp when the device data was received from the LWM2M Service          |

Reviewer 3: ```suggestion
| `timestampDataReceived` | Timestamp when the device data was received from the LWM2M service          |
```

| `sourceOfData`          | Describes the type of operation that triggered the data. This can be one of the following: |
|                       |  - `OBSERVATION_RESPONSE`: Data received in response to an observation request, triggered by the server. |
|                       |  - `NOTIFICATION`: Data received from the device as a notification, triggered by the device itself. |
|                       |  - `READ_RESPONSE`: Data received via any read operation, triggered by the server. |
|                       |  - `SEND`: Data received from the device via the LWM2M SEND command. |
| `deviceId`              | The device ID assigned upon registration in the platform                     |
| `tenantId`              | The ID of the tenant associated with the device                              |
| `lwm2mEndpoint`         | The device name as assigned to the platform                                  |
| `registrationId`        | The registration ID of the device                                            |
| `senMLMessage`          | Contains SenML-JSON formatted device measurement data, following the [RFC8428 specification](https://datatracker.ietf.org/doc/html/rfc8428) |

> Note that in `senMLMessage` LWM2M's Time resource data is always stored as a timestamp in seconds, 

Reviewer 3: ```suggestion
Note that in `senMLMessage` LWM2M's time resource data is always stored as a timestamp in seconds, 
```

> and the Base Time data is represented in scientific notation, which may include also fractional time information.

Reviewer 3: ```suggestion
and the base time data is represented in scientific notation, which may include also fractional time information.
```

---
weight: 100
title: Send Device Data to Cumulocity’s MQTT Service
layout: redirect
---

This guide helps you configure the {{< product-c8y-iot >}} to send device data to Cumulocity's MQTT Service. 
Once configured, you can easily access and process the data using your own microservice, giving you the flexibility 
to handle the data in a way that suits your specific needs.

{{< c8y-admon-important >}}
To ensure a smooth setup, please make sure that both Pulsar and the MQTT Service are properly configured and 
running before starting the {{< product-c8y-iot >}} configuration. Missing these prerequisites may result in excessive log 
entries indicating that the {{< product-c8y-iot >}} cannot establish a connection to the MQTT Service.
{{< /c8y-admon-important >}}

{{< c8y-admon-info >}}
If you are not familiar with MQTT, we recommend reviewing one of the many online resources available. For more 
details, you can visit the [MQTT website](https://mqtt.org/mqtt-specification/).
{{< /c8y-admon-info >}}

> ### What is MQTT Service?

Reviewer 2: In here we shouldn't really explain MQTT Service instead we can state that this is a required component for this feature.

Author: The idea of this section is to explain briefly the mqtt service and why we're using it. I can change the title to "Why MQTT Service?" and add some more arguments about this decision

The MQTT Service, developed by Cumulocity, offers several important benefits:

- **Multi-tenancy support**: A single endpoint serves multiple tenants, simplifying management.
- **Identity management**: Services connected to the MQTT Service use Cumulocity tenant platform credentials.
- **Efficient communication**: Low traffic overhead, ensuring efficient data transfer.
- **WebSocket and TLS support**: Secure and flexible communication options.
- **Scalability**: The service scales horizontally to meet your growing needs.
- **Custom payload formats**: The service uses SENML-JSON, a widely-used format for device measurement data.

Starting from release 2025, the {{< product-c8y-iot >}} can send device data to Cumulocity’s MQTT microservice, where the 
data is queued. This enables you to process and read the data via your custom service, with the added benefit 
of being able to leverage Cumulocity's integrated data analysis tools, should you choose to do so.

> ### Configuration Instructions

Reviewer 2: The configuration instructions should be in the OPS docs.
Users can't configure this.

You can instead say: To use this feature mqtt service options should be enabled for your tenant. Please contact support to enable.

Follow these steps to configure the {{< product-c8y-iot >}} to send device data to the MQTT Service:

| **Property** | **Description** |
|--------------|-----------------|
| `C8Y.mqtt.messaging.service.url` | Enter the WebSocket URL for the MQTT service (e.g., `ws://cumulocity.default.svc.cluster.local`). |
| `C8Y.mqtt.messaging.service.enabledForTenants` | List the tenants whose device data should be sent to the MQTT Service, separated by commas. |

If you provide the MQTT service URL but do not specify any tenants, 
the {{< product-c8y-iot >}} will not establish a connection.

> To retrieve data from the MQTT Service, you can deploy a custom microservice using the Microservice Deployer. 

Reviewer 2: microservice deployer also is not exposed to customers.

Author: If microservice deployer is not exposed to the customers, then how do they deploy their own services?

Reviewer 2: customer uploads from Administration microservice upload UI

The microservice will use the tenant’s credentials and must have the following roles assigned to function correctly:

- `ROLE_INVENTORY_READ`
- `ROLE_NOTIFICATION_2_ADMIN`
- `ROLE_MQTT_SERVICE_ADMIN`

Once deployed, the microservice will subscribe to the `lwm2m/data` topic, where the {{< product-c8y-iot >}} 
sends the device data. Note that it will only retrieve data for the specified tenant and its subtenants. 
If you need data from another tenant, you’ll need to deploy a separate microservice for that tenant.

For a detailed example of how to develop your own microservice, check out this 
[sample project](https://github.com/Cumulocity-IoT/lwm2m-devicedata-listener).

### MQTT Data Payload

The MQTT data payload uses the SENML-JSON format and follows the 
[RFC8428 specification](https://datatracker.ietf.org/doc/html/rfc8428). Keep in mind that 
LWM2M’s Time resource data is stored as a timestamp in seconds, while the Base Time may include 

> fractional data and is represented in scientific notation.

Reviewer 2: Could you give the example data format with topic: lwm2m/data here?

Author: sure
